pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/memory/CompoundDataBuffer.java

```diff
@@ -0,0 +1,672 @@
+      }
+      if (buffers[i].order() != _order) {
+        buffers[i] = buffers[i].view(0, buffers[i].size(), _order);
+      }
+    }
+    for (int i = 0; i < buffers.length; i++) {
+      _bufferOffsets[i] = offset;
+      long size = buffers[i].size();
+      offset += size;
+    }
+    _size = offset;
+  }
+
+  public CompoundDataBuffer(ByteBuffer[] buffers, ByteOrder order, boolean owner) {
+    this(asDataBufferArray(buffers), order, owner);
+  }
+
+  /**
+   * Creates a compound buffer from the given buffers.
+   * @param buffers The buffers that will be concatenated to form the compound buffer.
+   * @param order The byte order of the buffer. Buffers in the list that have a different byte order will be converted.
+   * @param owner Whether this buffer owns the underlying buffers. If true, the underlying buffers will be released when
+   *              this buffer is closed.
+   */
+  public CompoundDataBuffer(List<DataBuffer> buffers, ByteOrder order, boolean owner) {
+    this(buffers.toArray(new DataBuffer[0]), order, owner);
+  }
+
+  private static DataBuffer[] asDataBufferArray(ByteBuffer[] buffers) {
+    DataBuffer[] result = new DataBuffer[buffers.length];
+    for (int i = 0; i < buffers.length; i++) {
+      result[i] = PinotByteBuffer.wrap(buffers[i]);
+    }
+    return result;
+  }
+
+  private int getBufferIndex(long offset) {
+    // this optimistically assumes that lookups are going to be in ascending order
+    // we don't care about concurrency here given that this is only used to speed up the lookup
+    int lastBufferIndex = _lastBufferIndex;
+    if (_bufferOffsets[lastBufferIndex] > offset) {
+      lastBufferIndex = 0;
+    }
+
+    for (int i = lastBufferIndex; i < _bufferOffsets.length; i++) {
+      if (offset < _bufferOffsets[i]) {
+        int result = i - 1;
+        _lastBufferIndex = result;
+        return result;
+      }
+    }
+    return _bufferOffsets.length - 1;
+  }
+
+  private ByteBuffer copy(long offset, int length) {
+    if (offset + length > _size) {
+      throw new BufferUnderflowException();
+    }
+    byte[] result = new byte[length];
+
+    int bufferIndex = getBufferIndex(offset);
+    long inBufferIndex = offset - _bufferOffsets[bufferIndex];
+
+    DataBuffer buffer = _buffers[bufferIndex];
+    int toCopy = (int) Math.min(length, buffer.size() - inBufferIndex);
+    buffer.copyTo(inBufferIndex, result, 0, toCopy);
+
+    int remaining = length - toCopy;
+    while (remaining > 0) {
+      bufferIndex++;
+      buffer = _buffers[bufferIndex];
+      toCopy = (int) Math.min(remaining, buffer.size());
+      buffer.copyTo(0, result, length - remaining, toCopy);
+
+      remaining -= toCopy;
+    }
+    return ByteBuffer.wrap(result)
+        .order(_order);
+  }
+
+  @Override
+  public void readFrom(long offset, byte[] input, int srcOffset, int size) {
+    if (offset + size > _size) {
+      throw new BufferUnderflowException();
```

Comment:
I think this should be `BufferOverflowException`, same for other write methods

Reply:
Yeah! I always think `readFrom` is a read operation when it is a write operation